features/mbedtls/targets/TARGET_NUVOTON/TARGET_M480/ecp/ecp_internal_alt.c

```diff
@@ -0,0 +1,835 @@
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_read_eccreg(&R->Y, (uint32_t *) CRPT->ECC_Y1, NU_ECC_BIGNUM_MAXWORD));
+    MBEDTLS_MPI_CHK(mbedtls_mpi_lset(&R->Z, 1));
+        
+cleanup:
+
+    mbedtls_mpi_free(&N_);
+    
+    return ret;
+}
+
+int mbedtls_internal_run_modop(mbedtls_mpi *r,
+                            const mbedtls_mpi *o1,
+                            const mbedtls_mpi *o2,
+                            const mbedtls_mpi *p,
+                            uint32_t pbits,
+                            uint32_t modop)
+{
+    if (r == NULL || 
+        o1 == NULL ||
+        o2 == NULL ||
+        p == NULL) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check o1/o2 are not negative */
+    if (mbedtls_mpi_cmp_int(o1, 0) < 0 ||
+        mbedtls_mpi_cmp_int(o2, 0) < 0) {
+        return MBEDTLS_ERR_MPI_NEGATIVE_VALUE;
+    }
+    
+    /* Check p is positive */
+    if (mbedtls_mpi_cmp_int(p, 0) <= 0) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    /* Check supported maximum key bits */
+    if (pbits > NU_ECC_MAXKEYBITS) {
+        return MBEDTLS_ERR_ECP_FEATURE_UNAVAILABLE;
+    }
+    
+    /* Check MODOP operations are legal */
+    if (modop != MODOP_DIV &&
+        modop != MODOP_MUL &&
+        modop != MODOP_ADD &&
+        modop != MODOP_SUB) {
+        return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
+    }
+    
+    int ret;
+    bool ecc_done;
+
+    mbedtls_mpi N_;
+    const mbedtls_mpi *Np;
+    
+    mbedtls_mpi_init(&N_);
+    
+    /* Use MBEDTLS_INTERNAL_MPI_NORM(Np, N1, N_, P) to get normalized MPI
+     *
+     * N_: Holds normalized MPI if the passed-in MPI N1 is not
+     * Np: Pointer to normalized MPI, which could be N1 or N_
+     */
+
+    if (modop == MODOP_MUL ||
+        modop == MODOP_ADD ||
+        modop == MODOP_SUB) {
+        MBEDTLS_INTERNAL_MPI_NORM(&Np, *o1, N_, *p);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_X1, NU_ECC_BIGNUM_MAXWORD));
+         MBEDTLS_INTERNAL_MPI_NORM(&Np, *o2, N_, *p);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_Y1, NU_ECC_BIGNUM_MAXWORD));
+    } else if (modop == MODOP_DIV) {
+        MBEDTLS_INTERNAL_MPI_NORM(&Np, *o2, N_, *p);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_X1, NU_ECC_BIGNUM_MAXWORD));
+        MBEDTLS_INTERNAL_MPI_NORM(&Np, *o1, N_, *p);
+        MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(Np, (uint32_t *) CRPT->ECC_Y1, NU_ECC_BIGNUM_MAXWORD));
+    } else {
+        MBEDTLS_MPI_CHK(MBEDTLS_ERR_ECP_BAD_INPUT_DATA);
+    }
+
+    MBEDTLS_MPI_CHK(mbedtls_internal_mpi_write_eccreg(p, (uint32_t *) CRPT->ECC_N, NU_ECC_BIGNUM_MAXWORD));
+
+    crypto_ecc_prestart();
+    CRPT->ECC_CTL = (pbits << CRPT_ECC_CTL_CURVEM_Pos) | (ECCOP_MODULE | modop) | CRPT_ECC_CTL_FSEL_Msk | CRPT_ECC_CTL_START_Msk;
+    ecc_done = crypto_ecc_wait();
+    
+    /* FIXME: Better error code for ECC accelerator error */
```

Comment:
An error code for HW acceleration has recently been introduced in https://github.com/ARMmbed/mbedtls/pull/1303, but it's not yet in Mbed OS.

Reply:
@hanno-arm OK. I leave it unchanged currently.